openfermioncirq/primitives/_ffft.py

```diff
@@ -0,0 +1,192 @@
+        \langle 01_2 \rvert F_0^\dagger \lvert 10_2 \rangle = {1 \over \sqrt{2}}
+
+    .. math::
+        \langle 11_2 \rvert F_0^\dagger \lvert 11_2 \rangle = -1 \, .
+    """
+
+    def __init__(self):
+        """Initializes :math:`F_0` gate."""
+        cirq.TwoQubitMatrixGate.__init__(
+            self,
+            np.array([[1,          0,         0,  0],
+                      [0, -2**(-0.5), 2**(-0.5),  0],
+                      [0,  2**(-0.5), 2**(-0.5),  0],
+                      [0,          0,         0, -1]]))
+
+    def _circuit_diagram_info_(self, args: cirq.CircuitDiagramInfoArgs
+                               ) -> cirq.CircuitDiagramInfo:
+        if args.use_unicode_characters:
+            symbols = 'F₀', 'F₀'
+        else:
+            symbols = 'F0', 'F0'
+        return cirq.CircuitDiagramInfo(wire_symbols=symbols)
+
+
+class _TwiddleGate(cirq.ZPowGate):
+    r"""Gate that introduces arbitrary FFT twiddle factors.
+
+    Realizes unitary gate :math:`\omega^{k\dagger}_n` that phases creation
+    operator :math:`a^\dagger_x` according to:
+
+    .. math::
+        \omega^{k\dagger}_n a^\dagger_x \omega^k_n =
+            e^{-2 \pi i {k \over n}} a^\dagger_x \, .
+
+    Under JWT representation this is realized by appropriately rotated pauli Z
+    gate acting on qubit x.
+    """
+    def __init__(self, k, n):
+        """Initializes Twiddle gate.
+
+        Args:
+            k: Nominator appearing in the exponent.
+            n: Denominator appearing in the exponent.
+        """
+        cirq.ZPowGate.__init__(self, exponent=-2 * k / n, global_shift=0)
+        self.k = k
+        self.n = n
+
+    def _circuit_diagram_info_(self, args: cirq.CircuitDiagramInfoArgs
+                               ) -> cirq.CircuitDiagramInfo:
+        if args.use_unicode_characters:
+            symbols = 'ω^%d_%d' % (self.k, self.n),
+        else:
+            symbols = 'w^%d_%d' % (self.k, self.n),
+        return cirq.CircuitDiagramInfo(wire_symbols=symbols)
+
+
+F0 = _F0Gate()
+
+
+def ffft(qubits: Sequence[cirq.QubitId]) -> cirq.OP_TREE:
+    r"""TODO"""
+    n = len(qubits)
+
+    if n == 0:
+        raise ValueError('Number of qubits is 0.')
+
+    if n == 1:
+        return []
+
+    if n == 2:
+        return F0(*qubits)
+
+    if n % 2 != 0:
+        raise ValueError('Number of qubits is not a power of 2.')
+
+    ny = 2
+    nx = n // ny
+    permutation = [(i % ny) * nx + (i // ny) for i in range(n)]
+
+    operations = list()
+
+    operations.append(_permute(qubits, permutation))
+
+    for y in range(ny):
+        operations.append(ffft(qubits[nx * y:nx * (y + 1)]))
+
+    operations.append(_permute(qubits, _inverse(permutation)))
+
+    for x in range(nx):
+        for y in range(1, ny):
+            operations.append(_TwiddleGate(x * y, n).on(qubits[ny * x + y]))
+        operations.append(ffft(qubits[ny * x:ny * (x + 1)]))
+
+    operations.append(_permute(qubits, permutation))
```

Comment:
It re-permutes? That doesn't seem right.

Reply:
Added documentation which explains that.